feat(linux)!: Support streaming through XDG portals and Pipewire

ReenigneArcher opened the pull request:
<!---
    NOTE: DO NOT DELETE ANY COMMENTS OR SECTIONS OF THIS TEMPLATE.
    THEY ARE IMPORTANT FOR THE MAINTAINERS.
    IT IS OKAY TO LEAVE SECTIONS EMPTY AND BOXES UNCHECKED IF THEY DO NOT APPLY TO YOUR PR.
--->

## Description
<!--- Please include a summary of the changes. --->
This PR replaces https://github.com/LizardByte/Sunshine/pull/2507 ... Thanks to @garnacho for the initial work on this!

This PR adds a new "grab" implementation to use XDG desktop portals and Pipewire. There is a breaking change that comes along with this implementation, which is that KMS capture must now be started from a separte sunshine service named `sunshine-kms`. The main `sunshine` service now does not use extra capabilities defined by setcap.

This feature should be considered beta for now. Known issues:
- Monitoring scaling is not handled properly
- Touch and Pen passthrough (inputs that are relative to the display) are not working
- Cursor display toggle is not implemented

TODO:
- [x] update homebrew formula, depends on https://github.com/Homebrew/homebrew-core/pull/254211
- [x] fix software encoder tests (XDG portal should not be primary capture method... or environment is missing something)
  - [x] AppImage (fail)
  - [x] FreeBSD (fail and hang)
- [x] update cmake build args to explicitly include `-DSUNSHINE_ENABLE_PORTAL` throughout
- [x] reduce/eliminate sonar warnings


### Screenshot
<!--- Include screenshots if the changes are UI-related. --->


### Issues Fixed or Closed
<!--- If this PR fixes or closes any issues, please list them here. --->
<!--- Close issue example: `- Closes #1` --->
<!--- Fix bug issue example: `- Fixes #2` --->
<!--- Resolve issue example: `- Resolves #3` --->
- Closes https://github.com/LizardByte/Sunshine/pull/2507
- Resolves https://github.com/LizardByte/Sunshine/issues/3953
- Closes https://github.com/LizardByte/Sunshine/issues/3327
- Closes https://github.com/LizardByte/Sunshine/issues/3252

#### Roadmap Issues
<!--- If this PR solves any roadmap issues (https://github.com/LizardByte/roadmap/issues), please list them here. --->
<!--- e.g. `- Closes https://github.com/LizardByte/roadmap/issues/1` --->


## Type of Change
<!--- Select the type of change your PR introduces. You can select multiple types. --->
- [x] **feat**: New feature (non-breaking change which adds functionality)
- [ ] **fix**: Bug fix (non-breaking change which fixes an issue)
- [ ] **docs**: Documentation only changes
- [ ] **style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semicolons, etc.)
- [ ] **refactor**: Code change that neither fixes a bug nor adds a feature
- [ ] **perf**: Code change that improves performance
- [ ] **test**: Adding missing tests or correcting existing tests
- [ ] **build**: Changes that affect the build system or external dependencies
- [ ] **ci**: Changes to CI configuration files and scripts
- [ ] **chore**: Other changes that don't modify src or test files
- [ ] **revert**: Reverts a previous commit
- [x] **BREAKING CHANGE**: Introduces a breaking change (can be combined with any type above)


## Checklist
<!--- Please check all applicable boxes. If a box does not apply, leave it unchecked. --->
- [x] Code follows the style guidelines of this project
- [x] Code has been self-reviewed
- [x] Code has been commented, particularly in hard-to-understand areas
- [x] Code docstring/documentation-blocks for new or existing methods/components have been added or updated
- [ ] Unit tests have been added or updated for any new or modified functionality

### AI Usage
<!--- Select the option that best describes AI usage in this PR (choose only one). --->
- [x] **None**: No AI tools were used in creating this PR
- [ ] **Light**: AI provided minor assistance (formatting, simple suggestions)
- [ ] **Moderate**: AI helped with code generation or debugging specific parts
- [ ] **Heavy**: AI generated most or all of the code changes

garnacho committed:
third-party: Add stubs for EGL DMA-BUF query APIs

This is not present in the glad version imported in Sunshine,
plug the holes so that these functions may be used.

garnacho committed:
feat(linux): Support streaming through XDG portals and Pipewire

Add a new portal "grab" implementation that will request the necessary
permissions over XDG portals and use Pipewire for video data streaming.
This supports DMA-Buf buffer exchange for hardware accelerated encoding
(VAAPI, nvenc), software encoding will only work with memory buffers.

ReenigneArcher committed:
style: fix lint errors

ReenigneArcher committed:
Update XDG portal dependencies and error message

Require GIO and PIPEWIRE packages for portal support and use SYSTEM include directories. Also, clarify the fatal error message order for missing dependencies.

ReenigneArcher committed:
Add Pipewire as a dependency for Linux builds

Pipewire has been added to the dependency lists for FreeBSD, Arch, Fedora, Homebrew, and Debian-based packaging and build scripts. This ensures proper support and integration for systems using Pipewire.

carson-katri commented on src/platform/linux/portalgrab.cpp:
```diff
@@ -0,0 +1,799 @@
+#define CURSOR_MODE_EMBEDDED 2
+
+#define PERSIST_FORGET 0
+#define PERSIST_WHILE_RUNNING 1
```
I found that this will popup the "Remote Control" permissions dialog several times when starting sunshine, and when connecting a client.

Changing this to `2` gives the option to "Remember This Selection", which worked at least for the runtime of the sunshine process.

```suggestion
#define PERSIST_WHILE_RUNNING 2
```

ReenigneArcher committed:
fix(portalgrab): change PERSIST_WHILE_RUNNING to 2

A value of 2 indicates "Remember this Selection".

Co-authored-by: Carson Katri <Carson.katri@gmail.com>

ReenigneArcher commented:
> I've been playing with this branch a bit and I can't seem to get the pipewire screen selection pop up to work on KDE Plasma. I know this was previously mentioned in the superceded pull request. After unsetting the required cap for KMS, the following occurs

I haven't done much with this branch other than rebasing and getting it up to our new linting standards.

Does it work if you keep the setcap enabled?

marcelopio commented:
> Error: Could not create session: GDBus.Error:org.freedesktop.DBus.Error.AccessDenied: Portal operation not allowed: Unable to open /proc/<pid>/root

I had this problem running with systemd. If I compile this branch an run in a terminal, everything works.

Traine9 committed:
feat(linux/portal): improve XDG portal capture for KDE and dual GPU systems (#4510)

Co-authored-by: Carlos Garnacho <carlosg@gnome.org>
Co-authored-by: Carson Katri <Carson.katri@gmail.com>
Co-authored-by: d.bondarev <d.bondarev@crm-onebox.com>

ReenigneArcher committed:
fix(xdg)!: remove default setcap on sunshine binaries

ReenigneArcher committed:
refactor: move xdg portal capture after x11

ReenigneArcher committed:
Fix install path for sunshine-kms binary

Corrects the installation path for sunshine-kms to use %{buildroot}%{_bindir} instead of %{_bindir}, ensuring proper packaging for RPM builds.

ReenigneArcher committed:
style: fix several sonar warnings

ReenigneArcher committed:
Refactor to use std::array for buffer and struct storage

Replaced raw arrays with std::array for type safety and clarity in buffer and struct allocations throughout portalgrab.cpp. Updated related code to use .data() and .size() where appropriate, improving maintainability and reducing potential for errors.

ReenigneArcher committed:
Refactor array usage and variable declarations in portalgrab.cpp

Changed C-style string constants to array declarations for REQUEST_PREFIX and SESSION_PREFIX. Updated function calls to use std::array::data() for params. Split spa_pod_frame variable declarations for clarity.

ReenigneArcher committed:
Replace g_strdup_printf with std::format for path generation

Updated path string construction in portalgrab.cpp to use std::format instead of g_strdup_printf, improving type safety and readability.

ReenigneArcher committed:
Refactor portalgrab.cpp for code clarity

Applied minor refactoring for improved code clarity and consistency: added [[maybe_unused]] to unused parameters, replaced C-style casts with static_cast, removed unnecessary semicolon, and simplified std::fill_n usage.

ReenigneArcher committed:
Refactor parameter assignment and improve type usage

Refactored parameter assignment to avoid post-increment in array indexing and improve clarity. Updated token check to use direct character comparison. Replaced enum class usage with 'using enum' for mem_type_e. Improved logging for unrecognized capture status by using std::to_underlying.

ReenigneArcher committed:
Refactor portalgrab.cpp for const correctness and C++ casts

Updated function parameters and local variables to use const qualifiers where appropriate. Replaced C-style casts with C++ static_cast for improved type safety and clarity. Minor code cleanups and style improvements.

ReenigneArcher committed:
Enable SUNSHINE_ENABLE_PORTAL in build configurations

Added the -DSUNSHINE_ENABLE_PORTAL=ON flag to FreeBSD CI, COPR spec, Flatpak, and linux_build.sh scripts to ensure the portal feature is enabled across all build targets. Also reordered some CMake flags for consistency.

psyke83 committed:
fix(linux): Split service unit into privileged and unprivileged variants (#4619)

psyke83 committed:
fix(linux/xdgportal): sync capture delay logic/logging with x11/kms (#4622)

psyke83 committed:
fix(linux/xdgportal): set variable rate capture (#4624)

psyke83 committed:
fix(linux): install vendor preset for sunshine-kms service (#4626)

ReenigneArcher committed:
Format mark_as_advanced call for readability

Reformats the mark_as_advanced call in FindSystemd.cmake to list each variable on a separate line, improving readability and maintainability.

psyke83 committed:
fix(linux/kmsgrab): deprioritize setcap error message for non-KMS Wayland users (#4635)
